sys/progress_bar/progress_bar.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2019 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_progress_bar
+ * @{
+ *
+ * @file
+ * @brief       Progress bar implementation
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <inttypes.h>
+#include <assert.h>
+
+#include "progress_bar.h"
+
+void progress_bar_update(progress_bar_t *progress_bar)
+{
+    assert(progress_bar->value < 101);
+
+    /* Hide cursor */
+    printf("\033[?25l");
+
+    /* Move cursor at the beginning of the line */
+    putchar('\r');
+
+    /* Display progress bar prefix if any */
+    if (progress_bar->prefix[0]) {
+        printf("%s", progress_bar->prefix);
+    }
+
+    printf(PROGRESS_BAR_PREFIX_CHARACTER);
+
+    /* Fully reprint the progress bar */
+    for (unsigned j = 0; j < PROGRESS_BAR_LENGTH; j++) {
+        if (100 * j < (uint16_t)(progress_bar->value * PROGRESS_BAR_LENGTH)) {
+            printf(PROGRESS_BAR_FULL_CHARACTER);
+        }
+        else {
+            printf(PROGRESS_BAR_EMPTY_CHARACTER);
+        }
+    }
+
+    printf(PROGRESS_BAR_SUFFIX_CHARACTER);
+
+    /* Display progress bar suffix if any */
+    if (progress_bar->suffix[0]) {
```

Comment:
this check is redundant, printf will printf will not print if the string is empty.

Reply:
I removed it here but kept the check on the prefix as otherwise some garbage is printed if nothing is set in the buffer.

Reply:
How is that possible? Is the structure allocated on the stack and not emptied?
Shouldn't `printf("%s", { \0, ... })` print nothing?

Reply:
Declaring the progress bar static works but not on the stack indeed.

Reply:
That might be because defined statically, it'll be zeroed. On the stack it has random contents unless cleared.